pkg/apis/admission/types.go

```diff
@@ -73,6 +73,10 @@ type AdmissionRequest struct {
 	// OldObject is the existing object. Only populated for UPDATE requests.
 	// +optional
 	OldObject runtime.Object
+	// DryRun indicates that modifications will definitely not be persisted for this request.
+	// Defaults to false.
+	// +optional
+	DryRun *bool
```

Comment:
Is it really safe to make this optional? Do we need to make a completely new version?

Reply:
it has to be optional in v1beta1, because it's a new field, and has to default to false, since we can't assume current webhooks handle dry run correctly.

the current plan (in service of pushing all admission webhooks to support dry run) is described in https://github.com/kubernetes/community/blob/master/keps/sig-api-machinery/0015-dry-run.md#admission-controllers:

> If dry-run is requested on a non-supported webhook, the request will be completely rejected, as a 400: Bad Request. This field will be defaulted to `true` and deprecated in v1, and completely removed in v2. All webhooks registered with v2 will be assumed to support dry run.

Reply:
Yes, of course if we add it to the existing object it has to be optional.

The comment I'm working on leaving on the other API object will explain my position.

Reply:
are you proposing to only add dryRun to a v1 AdmissionReview object, require webhooks to use the v1 format to support dryRun, and treat all that use the v1beta1 format as not supporting dry run?

upsides:
* simpler signal that the webhook knows about dryrun
* no `default false in v1beta1, default true in v1, remove dryRunnable in v2` path required

downsides:
* harder for webhooks to support pre-1.12 and 1.12+ kube (they'd need distinct webhook registration manifests)

Reply:
If we require admission webhooks to use the v1 AdmissionReview format to support dry run, then all admission webhooks would have to be recompiled, whereas, adding it to v1beta1 allows webhooks which don't have side effects already (most webhooks probably fit into this category) to just change their config to add ```dryRunnable: true```

Reply:
> harder for webhooks to support pre-1.12 and 1.12+ kube (they'd need distinct webhook registration manifests)

Why?

Reply:
> Also, if a cluster admin wants to use dry run but some of their webhooks don't support it, they could manually add dryRunnable: true to the webhooks themself if they are willing to accept the risk in doing so

Yeah, this is a foot-gun :)

Reply:
>> harder for webhooks to support pre-1.12 and 1.12+ kube (they'd need distinct webhook registration manifests)
>
> Why?

They'd need to request v1beta1 AdmissionReview objects pre-1.12, and v1 AdmissionReview objects in 1.12+. I actually thought we already had the requested AdmissionReview version in the webhook registration object, but it turns out we don't.

Thinking about that a little more, we'll likely want a list of AdmissionReview versions the webhook can accept. Each apiserver, when sending an AdmissionReview object, would find the first version in the webhook's support list it is able to send. That would let 1.12 and pre-1.12 servers both talk to the same webhook, as long as it supported both v1 and v1beta1 AdmissionReview objects.

Reply:
>> if a cluster admin wants to use dry run but some of their webhooks don't support it, they could manually add dryRunnable: true to the webhooks themself if they are willing to accept the risk in doing so
>
> Yeah, this is a foot-gun :)

I don't think it's catastrophic... admission already has to tolerate and recover from being called in cases where the object isn't actually persisted to storage (because of a conflict, etc). telling the admission plugin this is a dry-run operation is largely an optimization that lets it skip eventually reconciling away a side-effect from a non-persisted request.